Docs: Improve Argument Clinic tutorial preamble

erlend-aasland opened the pull request:
Align with Diátaxis principles:

- Provide a complete picture before the tutorial starts
- Ignore options and alternatives: move suggested follow-up reading to
  the end of the tutorial, in the form of `.. seealso::`

Also:

- Link to the C API docs for parsing arguments


<!-- readthedocs-preview cpython-previews start -->
----
:books: Documentation preview :books:: https://cpython-previews--107323.org.readthedocs.build/

<!-- readthedocs-preview cpython-previews end -->

erlend-aasland committed:
Docs: Improve Argument Clinic tutorial preamble

Align with Diátaxis principles:

- Provide a complete picture before the tutorial starts
- Ignore options and alternatives: move suggested follow-up reading to
  the end of the tutorial, in the form of `.. seealso::`

erlend-aasland commented on Doc/howto/clinic.rst:
```diff
@@ -162,28 +162,20 @@ For the sake of clarity, here's the terminology we'll use with Argument Clinic:
-and :ref:`clinic-howto-self-converter`.
-But we'll keep it simple for this walkthrough so you can learn.
-
-First, make sure you're working with a freshly updated checkout
-of the CPython trunk.
```
I don't think this brings much value in the tutorial. The tutorial is targeted at the core dev and contributor audience; working on an old checkout seems like a very little likely scenario, IMO.

ezio-melotti replied:
While they know, it might be useful as a reminder.  If being a few commits behind is not an issue, then the reminder is not essential and can be removed.

erlend-aasland replied:
I think the tutorial currently mixes two things:

1. using a specific example (pickle.Pickler.dump)
2. trying to teach general guidelines for how to apply clinic to CPython (if x, then don't do y, if z then you can do w, etc.)

I'd like the tutorial to be repeatable and to not drift into digression after digression. The sentence in question could be a part of a "How to apply Argument Clinic to CPython" recipe, instead of part of the tutorial.

erlend-aasland replied:
I think it'll be hard to align the tutorial according to Diátaxis principles until we've extracted the "how to apply Argument Clinic to CPython" bits from it. Perhaps we should do that first.

What do you think, @ezio-melotti?

erlend-aasland commented on Doc/howto/clinic.rst:
```diff
@@ -162,28 +162,20 @@ For the sake of clarity, here's the terminology we'll use with Argument Clinic:
 
-If the call to the :c:func:`!PyArg_Parse*` function uses any of the
-following format units...:
+Next, take a look at the format string of the :c:func:`!PyArg_Parse*` call.
+If it contains any of the following format units...:
```
We should consider removing this from the tutorial, in my opinion. Alternatives and information are tutorial anti-patterns.

ezio-melotti replied:
You could state what format units are relevant for the tutorial, and link to the advanced converters section for the others.  IOW state what works, rather than what doesn't work.

ezio-melotti commented on Doc/howto/clinic.rst:
```diff
@@ -162,28 +162,20 @@ For the sake of clarity, here's the terminology we'll use with Argument Clinic:
-of the CPython trunk.
-
-Next, find a Python builtin that calls either :c:func:`PyArg_ParseTuple`
+In this tutorial, you will convert a C function from having custom
+:ref:`argument parsing code <arg-parsing>`
+to using Argument Clinic to generate that boilerplate code.
```
I feel this sentence could be better, maybe something like:
```suggestion
In this tutorial, you will learn how to use Argument Clinic to generate
:ref:`argument parsing code <arg-parsing>` for a function that
was using custom code.
```

The current sentence uses the imperative form to tell you what to do, without telling you why, in a way that sounds a bit harsh.  "Converting a function" seems something an howto would cover, whereas "learn how to use AC (to convert functions)" seems a better fit for the tutorial.

In addition, using "we" ("*we will learn how*") sounds "friendlier" to me, even thought technically only the reader is learning. This might require a refactoring of the whole tutorial, even though I see there are some instances of "we" later.  I don't remember if diataxis has any preference regarding this.

erlend-aasland committed:
gh-107507: Replace 'The goals of Argument Clinic' with a summary

Instead of repeating the rationale and motivation of PEP-436,
reword the Abstract to provide a summary of the goals of AC.

erlend-aasland committed:
Pull in main

erlend-aasland committed:
Pull in main
